xnmt/models/translators.py

```diff
@@ -124,76 +125,131 @@ def shared_params(self):
-      final_states = [self.encoder.get_final_states()]
-    else:
-      final_states = self.encoder.get_final_states()
-    initial_states = []
-    for encoding, final_state in zip(encodings, final_states):
-      self.attender.init_sent(encoding)
-      ss = batchers.mark_as_batch([Vocab.SS] * src.batch_size()) if batchers.is_batched(src) else Vocab.SS
-      initial_states.append(self.decoder.initial_state(final_state, self.trg_embedder.embed(ss)))
-    return CompoundSeqExpression(initial_states)
-
-  def calc_loss(self, src, trg, loss_calculator):
+  def _encode_src(self, src: Union[batchers.Batch, sent.Sentence]):
     self.start_sent(src)
-    initial_states = self._encode_src(src)
-    # Calculate losses from multiple initial states
+    embeddings = self.src_embedder.embed_sent(src)
+    encoding = self.encoder.transduce(embeddings)
+    final_state = self.encoder.get_final_states()
+    self.attender.init_sent(encoding)
+    ss = batchers.mark_as_batch([Vocab.SS] * src.batch_size()) if batchers.is_batched(src) else Vocab.SS
+    initial_state = self.decoder.initial_state(final_state, self.trg_embedder.embed(ss))
+    return initial_state
+
+  def calc_nll(self, src: Union[batchers.Batch, sent.Sentence], trg: Union[batchers.Batch, sent.Sentence]) -> dy.Expression:
+    # Encode the sentence
+    initial_state = self._encode_src(src)
+
+    dec_state = initial_state
+    trg_mask = trg.mask if batchers.is_batched(trg) else None
     losses = []
-    for initial_state in initial_states:
-      model_loss = FactoredLossExpr()
-      model_loss.add_factored_loss_expr(loss_calculator.calc_loss(self, initial_state, src, trg))
-  
-      if self.global_fertility != 0:
-        masked_attn = self.attender.attention_vecs
-        if trg.mask is not None:
-          trg_mask = 1-(trg.mask.np_arr.transpose())
-          masked_attn = [dy.cmult(attn, dy.inputTensor(mask, batched=True)) for attn, mask in zip(masked_attn, trg_mask)]
-        model_loss.add_loss("fertility", self._global_fertility(masked_attn))
-      losses.append(model_loss)
-    try:
-      total_loss = FactoredLossExpr()
-      list(total_loss.add_factored_loss_expr(x) for x in losses)
-      return total_loss
-    finally:
-      self.losses = losses
-
-  def calc_loss_one_step(self, dec_state:AutoRegressiveDecoderState, ref_word:batchers.Batch, input_word:Optional[batchers.Batch]) \
-          -> Tuple[AutoRegressiveDecoderState,dy.Expression]:
-    if input_word is not None:
-      dec_state = self.decoder.add_input(dec_state, self.trg_embedder.embed(input_word))
-    rnn_output = dec_state.rnn_state.output()
-    dec_state.context = self.attender.calc_context(rnn_output)
-    word_loss = self.decoder.calc_loss(dec_state, ref_word)
-    return dec_state, word_loss
-
-  def generate(self, src: batchers.Batch, idx: Sequence[int], search_strategy: SearchStrategy, forced_trg_ids: batchers.Batch=None):
+    seq_len = trg.sent_len()
+
+    # # TODO: enable only when debugging
```

Comment:
maybe add a ```CONSISTENCY_CHECKS``` flag to the settings module which is true only for debug settings, and then use ```if settings.CONSISTENCY_CHECKS:```?

Reply:
We already have `CHECK_VALIDITY`, maybe I'll just reuse this.